### PR TITLE
Connect revenue report table to the API

### PR DESCRIPTION
**Pull request by emdashcodes**
This is a part of #225.

It connects the table part of the revenue report up to the API. You can page through the summary results and reorder columns. Other elements on the report are still hooked up to the swagger API. More to come later.

screen shot 2018-08-15 at 1 17 25 pm

To Test:
* run `npm test` and make sure all tests pass
* Before testing on your site, you may want to generate a number of orders. You can use [wc-smooth-generator](https://github.com/woocommerce/wc-smooth-generator) to do this. A PR of mine was just merged that allows you some control over date ranges for the generated orders. See the [README](https://github.com/woocommerce/wc-smooth-generator/blob/master/README.md#orders) for an example.
* Pull the latest WooCommerce v3 API from `feature/rest-api-v3-reports` (you might need to deactivate `wc-api-dev`).
* You need to deactivate and reactivate WooCommerce to get the correct summary MySQL tables. There is also a tool under system settings to rebuild the summary tables.
* Visit the revenue report.
* Test reordering columns, paging results, etc.

### CODE REVIEW

**Comment by ryelle** [comment]
Hm, I can't seem to test this because I keep getting `PHP Notice:  Trying to get property of non-object in woocommerce/includes/api/class-wc-rest-reports-revenue-stats-controller.php on line 81`

We'll probably want some better designed error state, but that's an issue for another PR

screen shot 2018-08-15 at 4 29 05 pm

File: client/components/table/placeholder.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+									);
+								} ) }
+							</tr>
+							{ range( rows ).map( i => {
+								return (
+									<tr key={ i }>
+										{ headers.map( ( header, j ) => {
+											const Cell = 0 === j ? 'th' : 'td';
+											const cellClasses = classnames( 'woocommerce-table__item' );
+											return <Cell scope="row" key={ j } className={ cellClasses } />;
```
Comment (ryelle) [inline client/components/table/placeholder.js]: Since there's no real semantics gained from `th`/`td` on empty rows, and no conditional classnames, you could shorten this to 

```jsx
headers.map( ( header, j ) =>  <td key={ j } className="woocommerce-table__item" /> )
```

(also because having the row scope on every cell is meaningless, and apparently something we should fix in `<Table />` - it should only be applied to the `th`)

**Comment by justinshreve** [comment]
ping @woocommerce/woo-core-design re error states. What do we want to do/display if we have trouble loading report information from the API? As you can see above, I have a temporary message but we can update that in another PR.

**Comment by ryelle** [comment]
@justinshreve Well, I didn't change anything but it's working now -- maybe the rebuilding took extra-long for some reason. I'll check this out this morning.

File: client/analytics/style.scss [hunk 2]
```diff
@@ -0,0 +1,14 @@
+/** @format */
+
+.woocommerce-analytics__report-table-card {
+	&.is-requesting {
+		animation: loading-fade 1.6s ease-in-out infinite;
```
Comment (ryelle) [inline client/analytics/style.scss]: Could you use the `placeholder` mixin here?

File: client/analytics/report/revenue.js [hunk 3]
```diff
@@ -212,13 +220,68 @@ class RevenueReport extends Component {
...
 		const rows = this.getRowsContent( intervals ) || [];
-		const headers = this.getHeadersContent();
 
+		const className = classnames( 'woocommerce-analytics__report-table-card', {
+			'is-requesting': isTableDataRequesting,
+		} );
+
+		return (
+			<TableCard
+				title={ __( 'Revenue Last Week', 'wc-admin' ) }
```
Comment (ryelle) [inline client/analytics/report/revenue.js]: This seems to be fetching all of my report data, not just within the last week. The pagination & column sorting approach we've taken assume an "all revenue data (per day)" functionality, so I think we just need to relabel this table. Thoughts @woocommerce/woo-core-design  ?

**Review by timmyc — COMMENTED** [review]
AWESOME! So much great stuff here Justin - first off I had to completely setup my local dev stack to test due to vvv being odd on me, so your pointers about smooth generator made that a snap.

Interacting with the table was pure `:whoa:` - everything is working as expected... the only minor oddity i found i noted around the double `//` in the URL for the orders link. Just so awesome to see all the components wired into real data - such a big milestone!

Some comments/questions in the code, but most are me just trying to learn a bit more from what you have discovered in your `wp.data` journeys.

Would you like for me to create new issues for the TODO's in the code here?

File: client/analytics/report/revenue.js [hunk 4]
```diff
@@ -212,13 +220,68 @@ class RevenueReport extends Component {
...
+				query={ query }
+				summary={ summary }
+				className={ className }
+			/>
+		);
+	}
+
+	render() {
+		const { path, query } = this.props;
+		const { totals = {} } = this.state.stats;
```
Comment (ryelle) [inline client/analytics/report/revenue.js]: This `totals` is still fake data because we need to query for the table by week and _also_ query the previous week to get the deltas, right?

That kind of messes with my previous comment about how the table has been developed to _not_ use week queries, hm … 🤔

File: client/components/table/style.scss [hunk 5]
```diff
@@ -10,6 +10,11 @@
 	.woocommerce-card__menu {
 		justify-self: flex-end;
 	}
+
+	&.is-placeholder {
+		animation: loading-fade 1.6s ease-in-out infinite;
```
Comment (ryelle) [inline client/components/table/style.scss]: Do you need the styles in `analytics/style.scss` if you're also styling the placeholder here?

**Review by ryelle — COMMENTED** [review]
We should move revenue into its own folder like orders & products are (though none of these should need a style.css now that all the filter components have been merged). That can be done in a follow up PR, though.

I'm concerned about the table data & interactions (pagination/sorting) we've set up, since that assumes "view all data", but other UI elements assume "view a weeks worth of data". For example, if you sort a column, you're no longer looking at just that latest week, what does that do to the SummaryNumbers? If you change the pagination "per page", you're not looking at the last week - so should SummaryNumbers update to compare vs the previous "page" of data?

But that's more of a UX thing, code-wise this is looking good 👍

File: client/analytics/report/index.js [hunk 6]
```diff
@@ -14,17 +15,35 @@ import ProductsReport from './products';
...
 class Report extends Component {
+	constructor( props ) {
+		super( props );
+		this.state = {
+			previousQuery: props.query,
+		};
+	}
+
+	componentDidUpdate( prevProps ) {
+		// Track the previous query, so we can continue displaying results in reports while updated results are pulled.
```
[the remark this answers is not in the document]
Comment (justinshreve) [reply client/analytics/report/index.js]: Right now I have the content do a loading pulse as new results are fetched.

File: client/store/reports/revenue/stats/resolvers.js [hunk 7]
```diff
@@ -15,10 +15,13 @@ import { NAMESPACE } from 'store/constants';
 export default {
 	async getReportRevenueStats( state, query ) {
 		try {
-			const report = await apiFetch( {
+			const response = await apiFetch( {
 				path: NAMESPACE + 'reports/revenue/stats' + stringifyQuery( query ),
+				parse: false,
```
[the remark this answers is not in the document]
Comment (justinshreve) [reply client/store/reports/revenue/stats/resolvers.js]: `parse` does a few things like abstract the body out of the response. This allows me to get at the headers.

File: client/analytics/report/revenue.js (hunk 3 above)
Comment (justinshreve) [reply client/analytics/report/revenue.js]: Yeah good catch. This table is not just for the last week - it lets you page through all results. Should just this just be "Revenue"?

**Comment by justinshreve** [comment]
> I'm concerned about the table data & interactions (pagination/sorting) we've set up, since that assumes "view all data", but other UI elements assume "view a weeks worth of data". For example, if you sort a column, you're no longer looking at just that latest week, what does that do to the SummaryNumbers? If you change the pagination "per page", you're not looking at the last week - so should SummaryNumbers update to compare vs the previous "page" of data?

> But that's more of a UX thing,

@ryelle 

It was my understanding that the table is independent from the SummaryNumbers and chart.

So the table would always be "per day", and allow you to page/filter through all results. The sorting would only affect the table queries. The table would not use the date filtering.

Charts and Summary Numbers would use the date filtering.

@woocommerce/woo-core-design can verify or clarify here, but I think we are OK with the UX.

File: client/analytics/report/revenue.js (hunk 4 above)
Comment (justinshreve) [reply client/analytics/report/revenue.js]: Yes. I left this using fake data for now, because we will need different queries for the summary numbers/chart (including to get the delta).

**Comment by ryelle** [comment]
> It was my understanding that the table is independent from the SummaryNumbers and chart.

I was looking at p7bje6-14c-p2, which makes it sound like we would use the same API requests to get the data for tables & summary numbers, but now that I look at it again I think I read it wrong - we'll use `interval=day` for the table & graph, and `interval=week` for the SummaryNumbers. It _does_ sound like the same request would be used for the table and graph though, but I think we've changed that since the post was written. So we'll have these queries for the report: 

- revenue per week for current period + previous period to get the SummaryNumbers & deltas
- revenue per day for current period + previous period to create the graph
- X latest days of revenue info to create the chart, which can be paginated or sorted, and pagination+sorting of the table will not change the graph or SummaryNumbers

If that's right, then we're on the same page, I think that p2 post just threw me a bit.

**Comment by jameskoster** [comment]
For the error, can we use our Empty State component with a button to reload the page?

I'll make a todo to create a graphic. Will get to that next week.

**Comment by jameskoster** [comment]
For table, the idea was that the summary numbers / charts were built from the contents of that table. So _everything_ in the report should be governed by the date-picker. 

> since that [the table] assumes "view all data", but other UI elements [summary numbers & charts] assume "view a weeks worth of data".

If the table also assumes "view a weeks worth of data" does that solve the sorting issues?

**Comment by justinshreve** [comment]
Thanks for the reviews and feedback, everyone!

Per https://a8c.slack.com/archives/C9K5T0XSP/p1534526650000100, I have a better understanding of how all the elements are supposed to work together. I'm going to close this PR out and work on integrating it with a chart branch I already started. I'll also take the feedback above into account.
